### PR TITLE
Plot Waterfall Components Decomposition

### PR DESCRIPTION
**Pull request by cetagostini-wise**
## Related Issue

- [x] Closes #625 
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Modules affected

- [x] MMM
- [ ] CLV

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [x] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pymc-marketing--631.org.readthedocs.build/en/631/

### CODE REVIEW

**Comment by cetagostini** [comment]
Any idea why `PyPI / build source distribution (pull_request) Failing after 22s` is failing? @wd60622 @juanitorduz

**Comment by juanitorduz** [comment]
> Any idea why `PyPI / build source distribution (pull_request)` is failing? @wd60622 @juanitorduz

I think it is related to https://github.com/pymc-devs/pymc-experimental/pull/329#issuecomment-2056646766  and https://github.com/pymc-devs/pymc-experimental/commit/1f70c7afc22328fb761e727ad5281930d6a3db87 I will create a PR

**Comment by juanitorduz** [comment]
i think https://github.com/pymc-labs/pymc-marketing/pull/633 solves it :D

File: pymc_marketing/mmm/base.py [hunk 1]
```diff
@@ -1293,6 +1295,130 @@ def plot_channel_contribution_share_hdi(
     def graphviz(self, **kwargs):
         return pm.model_to_graphviz(self.model, **kwargs)
 
+    def _process_decomposition_components(self, data: pd.DataFrame):
+        """
+        Process data to compute the sum of contributions by component and calculate their percentages.
+        Parameters
```
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:1301]: Can you leave a blank line before "Parameters"?

File: pymc_marketing/mmm/base.py [hunk 2]
```diff
@@ -1293,6 +1295,130 @@ def plot_channel_contribution_share_hdi(
     def graphviz(self, **kwargs):
         return pm.model_to_graphviz(self.model, **kwargs)
 
+    def _process_decomposition_components(self, data: pd.DataFrame):
```
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:1298]: return type hints missing

File: pymc_marketing/mmm/base.py [hunk 3]
```diff
@@ -1293,6 +1295,130 @@ def plot_channel_contribution_share_hdi(
...
+        """
+        Process data to compute the sum of contributions by component and calculate their percentages.
+        Parameters
+        ----------
+        dataframe : pd.DataFrame
+            Dataframe containing the time series data.
+        Returns
```
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:1305]: Can you leave a blank line before "Returns"?

File: pymc_marketing/mmm/base.py [hunk 4]
```diff
@@ -1293,6 +1295,130 @@ def plot_channel_contribution_share_hdi(
     def graphviz(self, **kwargs):
         return pm.model_to_graphviz(self.model, **kwargs)
 
+    def _process_decomposition_components(self, data: pd.DataFrame):
+        """
+        Process data to compute the sum of contributions by component and calculate their percentages.
+        Parameters
+        ----------
+        dataframe : pd.DataFrame
+            Dataframe containing the time series data.
```
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:1304]: Can you please more explicit about the columns and dimensions expected by this function? Think of it from a user perspective :)

File: pymc_marketing/mmm/base.py [hunk 5]
```diff
@@ -1293,6 +1295,130 @@ def plot_channel_contribution_share_hdi(
...
+
+        total_contribution = dataframe["contribution"].sum()
+        dataframe["percentage"] = (dataframe["contribution"] / total_contribution) * 100
+
+        return dataframe
+
+    def plot_waterfall_components_decomposition(
+        self,
+        original_scale: bool = True,
+        figsize: Tuple = (14, 7),
```
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:1326]: can you use `Tuple[int, int]` to be more explicit :)

File: pymc_marketing/mmm/base.py [hunk 6]
```diff
@@ -1293,6 +1295,130 @@ def plot_channel_contribution_share_hdi(
...
+
+        # Sort the dataframe in ascending order of contribution for the waterfall plot
+        dataframe = self.compute_mean_contributions_over_time(
+            original_scale=original_scale
+        )
+
+        dataframe = self._process_decomposition_components(data=dataframe)
+        total_contribution = dataframe["contribution"].sum()
+
+        # Initialize the matplotlib figure and axis
```
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:1355]: you can remove this comment :)

File: pymc_marketing/mmm/base.py [hunk 7]
```diff
@@ -1293,6 +1295,130 @@ def plot_channel_contribution_share_hdi(
...
+        # Hide the right, top, and left spines for a cleaner look
+        ax.spines["right"].set_visible(False)
+        ax.spines["top"].set_visible(False)
+        ax.spines["left"].set_visible(False)
+
+        # Add labels on the left to identify the predictor channels, corresponding to the y-ticks
+        ax.set_yticks(np.arange(len(dataframe)))
+        ax.set_yticklabels(dataframe["component"])
+
+        plt.tight_layout()
```
Comment (juanitorduz) [inline pymc_marketing/mmm/base.py:1419]: i think you can remove this  (I think it has been depreciates but I am not sure) and use 

```python
fig, ax = plt.subplots(figsize=figsize, layout-"constrained", **kwargs)
```

File: tests/mmm/test_delayed_saturated_mmm.py [hunk 8]
```diff
@@ -286,7 +286,7 @@ def test_fit(self, toy_X: pd.DataFrame, toy_y: pd.Series) -> None:
         assert mmm.model_config is not None
         n_channel: int = len(mmm.channel_columns)
         n_control: int = len(mmm.control_columns)
-        fourier_terms: int = 2 * mmm.yearly_seasonality
+        # fourier_terms: int = 2 * mmm.yearly_seasonality
```
Comment (juanitorduz) [inline tests/mmm/test_delayed_saturated_mmm.py:289]: is this intended? Either uncomment or remove the line :)

**Comment by juanitorduz** [comment]
> @juanitorduz Perfect, I'll make the change!

No need to! I will merge the fix and then you can update the branch from this PR very easily with the new "green button" :)

**Review by juanitorduz — CHANGES_REQUESTED** [review]
@cetagostini-wise Nice PR and easy to review! I added some small comments but overall looks great!

**Comment by juanitorduz** [comment]
@cetagostini you can not push the "Update Branch" buttton ⬆️

File: pymc_marketing/mmm/base.py (hunk 5 above)
Comment (juanitorduz) [reply pymc_marketing/mmm/base.py:1327]: Actually, after https://github.com/pymc-labs/pymc-marketing/pull/635  we should use `tuple`

Comment (cetagostini-wise) [reply pymc_marketing/mmm/base.py:1333]: Change!

File: pymc_marketing/mmm/base.py (hunk 4 above)
Comment (cetagostini) [reply pymc_marketing/mmm/base.py:1311]: Added

**Comment by cetagostini** [comment]
Ready for review @juanitorduz

**Review by juanitorduz — APPROVED** [review]

**Comment by juanitorduz** [comment]
Thank you @cetagostini !

**Comment by juanitorduz** [comment]
@cetagostini do you want to do a following PR where you add this plot to the mmm example notebook?

**Comment by cetagostini** [comment]
> @cetagostini do you want to do a following PR where you add this plot to the mmm example notebook?

Sure, I'll make it @juanitorduz

**Pull request merged**